# Tasks Assessment

This notebook will hold my solutions to the Tasks Assessment for Machine Learning and Statistics 2020.


The author is Angela Carpenter

# Table of Contents

- [Task 1: Calculate the square root of two to 100 decimal places without using any imported module](#task-1)


- [Task 2: Chi-squared test for Independence](#task-2)

- [Task 3 - Standard Deviation Functions](#task-3)

- [Task 4 - not yet released](#task-4)

- [References](#references)
***

#### Plan
- Cut more from task 1! 
- Add some more references for task 2 rather than just referring old college notes.
- alpha level, p-value etc
- Tidy up references, keep numbering consistent.
- Task 3 Standard deviation 
    


## Task 1
Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code should not depend on any module from the standard library or otherwise. You should research the task first and include references and a description of your algorithm.

## The square root of two $\sqrt{2}$
In mathematics, a square root of a number $x$ is a number $y$ such that $y2 = x$; in other words, a number $y$ whose square (the result of multiplying the number by itself, or $y⋅y$) is $x$. [1][Wikipedia](https://en.wikipedia.org/wiki/Square_root).
The principal square root function $f(x) = \sqrt{x}$ is a function that maps the set of nonnegative real numbers onto itself. In geometrical terms, the square root function maps the area of a square to its side length.

The square root of two is the number that when multiplied by itself gives the number two. The length of the diagonal of a square where each side of the square is of length 1 is the square root of the number 2 (Pythagoras theorem).

According to [2][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2)., the square root of two was probably the first number known to be irrational.
>*The square root of 2, or the one-half power of 2, written in mathematics as 
$2{\sqrt {2}}$ or $2^{{1/2}}$ is the positive algebraic number that, when multiplied by itself, equals the number 2.*
>Technically, it must be called the principal square root of 2, to distinguish it from the negative number with the same property.
[2][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2).


An **irrational** number is one which cannot be expressed as a ratio of two numbers, i.e. a fraction.
Any number that can be expressed as a ratio of two integers, a decimal that either ends or repeats in a recurring manner or a percentage is said to be a rational number. All whole numbers are rational as they can be represented as fractions divided by 1. 

Greek mathematicians such as Pythagoras had believed that all numbers could be shown to be fractions, believing that the number line was made up entirely of fractions, because for any two fractions we can always find another fraction in between them. Because this process has no end there are infinitely many such points. 
Hippasus [4][The Maths Book: Big Ideas Simply explained](published by Dorling Kindersley Limited), a Greek scholar is believed to have first identified irrational numbers in the 5th century BCE as he worked on geometrical problems. He applied Pythagoras's theorem which states that the square of the hypothenuse in a triangle is equal to the sum of the squares of the other two sides, to a right-angled triangle with both shorter sides equal to one. Hippasus realised that the length of the hypothesus of such as triangle was equal to the square root of two but that the square root of two could not be expressed as a ratio of two whole numbers, as a fraction. There is no rational number which can be multiplied by itself to give precisely two which means that the square root of two an irrational number. The number two is termed non-square or square free. The numbers 3, 5 and 7 and many others are also non-square with irrational square roots. Numbers such as 4, 9 and 16 are square numbers and do have square roots that are also whole numbers and therefore rational.

Another rational number can always be found in between two rational numbers. The average of the two numbers will also be rational. The average of the number and either of the two original numbers will be rational. Irrational numbers can also be found in between any two rational numbers. There are infinite numbers of rational numbers and infinite numbers of irrational numbers but the 

A real number gives a positive result when squared. The square root of two gives a positive result when squared of 2. An irrational number has an infinite number of decimals with no recurring pattern. This is the case for the square root of two.  The square root of 2 truncuated to 50 decimal places is 1.41421356237309504880168872420969807856967187537694
See (https://en.wikipedia.org/wiki/Square_root).The square root of 2 is a real number that is not rational.


The square root of $x$ is **rational** if and only if $x$ is a rational number that can be represented as a ratio of two perfect squares.
A positive number has two square roots, one positive and one negative which are opposite each other but when talking of the square root of a positive integer it is usually just the positive square root that is meant.
The square root of a positive integer is the product of the roots of its prime factors, because the square root of a product is the product of the square roots of the factors.

The square roots of perfect squares such as 1, 4, 9 and 16 are integers. In all other cases, the square roots of positive integers are irrational numbers, and therefore have non-repeating digits in any standard positional notation system.

The value of the square root of two cannot be determined accurately because it cannot be represented as a fraction in the form a/b where a and b are whole numbers and has an infinite number of decimals.[3][geeksforgeeks](https://www.geeksforgeeks.org/square-root-of-2/)

In mathematics, the natural numbers are those used for counting and ordering. Some definitions begin the natural numbers at 0 corresponding to non-negative integers 0,1,2,3... collectively denoted by $\mathbb{N_0}$  while others start with 1, corresponding to the positive integers 1,2,3,... collectively denoted by $\mathbb{N_1}$. The natural numbers are the basis from which other number sets are built by extension. Natural numbers are often known as the counting numbers (excluding negative integers and zero) and are discrete numbers in contrast to measuring numbers or real numbers where the values are continuous and represent a distance along a line. Real numbers can be represented as an inifinite decimal expansion.

####  Binary representation of integers and floats and what this means for the square root of two.
In terms of computing, computer programs are made up of binary strings of 1's and 0's which can be converted into integers. Every computer program can be converted into one of the counting numbers from the set of $\mathbb{N}$ . According to Turing, all computer programs are countable which means they can be put into one-to-one correspondance with the natural numbers but Turing also realised that not all numbers can be calculated or computed.[Lectures notes by Ian Mc Loughlin]. The set of $\mathbb{N}$ Natural numbers are an infinite set but it is the smallest infinite set there is and is known as countable infinite. The set of measuring numbers are all the numbers on the number line that are used for quantifying and measuring things and there is an infinite amount of such numbers between any two discrete points on the number line.

A rational number can be expressed as the ratio of 2 whole numbers. It can be stored as an array of two numbers. (A number is rational only if it's decimal expansion becomes periodic).
Real numbers such as $\sqrt{2}$ have an infinite number of places after the decimal point. Computers cannot really store such real numbers as there is only a finite number of bits available to store the number.

Floating point numbers are stored on a computer with a finite number of digits so there is a finite limit on how many decimal places they can actually store. Floating point numbers are actually stored using integers in two parts, for the whole number part before the decimal point (the characteristics) and the part of the whole after the decimal point (the mantissa). 

The square root of two is an irrational number and therefore cannot be expressed in the ratio of two numbers in their most simplified form with the resulting implications being that the square root of two can never be stored in a computer. Instead an approximation of its value can be stored (in the same way an approximation for the irrational number $pi$ is stored). Irrational numbers are often stored as procedures which are used to calculate but not store the number.

For this reason caution is always required when calculation statistics or performing analytics on irrational numbers. It is possible to work theoretically or to use approximations but rounding errors could interfere will the calculations.

[Chapter 5 of the Python Tutorial](https://docs.python.org/3/tutorial/floatingpoint.html) looks at the issues and limitations associated with floating point arithmetic. It illustrates how the decimal fraction 0.125 and the binary fraction 0.001 have identical values:

- Floating-point numbers are represented in computer hardware as base 2 (binary) fractions
- The decimal fraction 0.125 has value $\frac{1}{10}$ + $\frac{2}{100}$ + $\frac{5}{1000}$
- The binary fraction 0.001 has value $\frac{0}{2}$ + $\frac{0}{4}$ + $\frac{1}{8}$
The only real difference here is that the decimal fraction 0.125 is written in base 10 fractional notation while the binary fraction 0.001 above is written in base 2 notation.

>Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.

(It demonstrates how the fraction 1/3 can be approximated as a base 10 fraction as  0.3 or 0.33 or 0.333. No matter how many 3's you use the result will never actually be $\frac{1}{3}$ but it will be increasingly a better approximation of $\frac{1}{3}$)

The decimal value 0.1 cannot be represented exactly as a base 2 fraction because in base 2, $\frac{1}{10}$ is the infinitely repeating fraction
0.0001100110011001100110011001100110011001100110011...
If you stop at any  finite number of bits you get an approximation. 
On most machines today floats are approximated using a binary fraction. 
- The numerator uses the first 53 bits starting with the most significant bit
- The denominator as a power of 2
The binary fraction for the decimal 0.1 is `3602879701896397 / 2 ** 55` which is close to the true value of 1/10 but it is not exactly equal to the true value of 0.1.
We don't notice that this is not the true value and only an approximation because of the way values are displayed. Python (and other languages) only prints *a decimal approximation to the to the true decimal value of the binary approximation stored by the machine* 

- The true decimal value of the binary approximation stored for 0.1 is actually `0.1000000000000000055511151231257827021181583404541015625`. It is printed as 0.1 though.
- Python displays a rounded value instead of all the digits as in most situations this is enough.
- The important point is that even though what we see printed (0.01) looks like the exact value of $\frac{1}{10}$, the actual value that is stored is the nearest representable binary fraction.

- The Python docs tutorial also illustrates how there are many different decimal numbers which actually share the same nearest approximate binary fraction. The numbers `0.1`  and `0.10000000000000001` and `0.1000000000000000055511151231257827021181583404541015625` are all approximated by `3602879701896397 / 2 ** 55` . These decimal values all have the same approximation.

- String formatting can be used to produce a limited number of significant digits.
For example `format(math.pi, '.12g')` gives 12 significant digits while `format(math.pi, '.2f')` gives 2  digits after the point.
However this is simply rounding the **display** of the true machine value.

See [The Perils of Floating Point](http://www.lahey.com/float.htm)
See [str.format](https://docs.python.org/3/library/stdtypes.html#str.format)

The errors in Python float operations come from the floating-point hardware. 
>The errors in Python float operations are inherited from the floating-point hardware, and on most machines are on the order of no more than 1 part in `2**53` per operation. That’s more than adequate for most tasks, but you do need to keep in mind that it’s not decimal arithmetic and that every float operation can suffer a new rounding error. (Chapter 5)

**Representation error** refers to the fact that most decimal fractions cannot be represented exactly as binary (base 2) fractions in Python and other languages and therefore won’t display the exact decimal number you expect.
- Most machines use IEEE-754 floating point arithmetic.
- Almost all platforms map Python floats to IEEE-754 “double precision”.
- 754 doubles contain 53 bits of precision. 
The computer strives to convert 0.1 to the closest fraction that it can of the form $J/2**N$ where J is an integer containing exactly 53 bits.

- See [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) and [IEEE Standard 754 Floating Point Numbers](https://www.geeksforgeeks.org/ieee-standard-754-floating-point-numbers/) article on <geeksforgeeks.com>.
- IEEE 754 has 3 basic components:
    * The sign of the mantissa where `0` represents a positive number while `1` represents a negative number.
    * The biased exponent
    * The Normalised Mantissa

- The [decimal module](https://docs.python.org/3/library/decimal.html#module-decimal) implements decimal arithmetic suitable for accounting applications and high-precision applications.
- The [fractions module](https://docs.python.org/3/library/fractions.html#module-fractions)  implements arithmetic based on rational numbers (so the numbers like 1/3 can be represented exactly).
- NumPy and [scipy](https://scipy.org) packages are recommended for mathematical and statistical operations.
- Python also has a [float.as_integer_ratio()](https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio) method which expresses the value of a float as a fraction. This can help when you want to know the exact value as a float. As the ratio is exact, it can be used to losslessly recreate the original value.


---

Both Python's `math` module and the `numpy` library have a function called `sqrt` for calulating square roots.
This task should not use any libraries but they can be used to check the code.
`math.sqrt(x)` calculates the square root of whatever is passed in as `x`.
This task however is to calculate the square root of the number 2 to 100 decimal places. 
The square root of a number when multiplied by itself should give the number so here I check that by multiplying the result of `math.sqrt(2)` by `math.sqrt(2)`.

The results below of squaring the square root of two returned from both the `math.sqrt` function and the `np.sqrt` function is 2.0000000000000004.  This demonstrates the problem with performing calculations on the irrational number square root of two!


will delete all this below but leaving here for my reference for now.

In [233]:
# import math module
import math
# using the sqrt function
print(math.sqrt(2))
print(math.sqrt(2)*math.sqrt(2))
import numpy as np
print(np.sqrt(2))
print(np.sqrt(2)*np.sqrt(2))
np.sqrt(2)*np.sqrt(2)==2
# just seeing if the numpy and math square roots are the exact same
math.sqrt(2) == np.sqrt(2)
math.sqrt(2)*math.sqrt(2) == np.sqrt(2)*np.sqrt(2)
x = np.sqrt(2)
x.as_integer_ratio()

1.4142135623730951
2.0000000000000004
1.4142135623730951
2.0000000000000004


True

## Methods for Calculating square roots
Finding the square root of a number is the inverse of finding the square of a number. The square of a number is the number times itself. The perfect squares are the squares of the whole numbers. The square root of a number is the number that when multiplied by itself gives the number.

>*The square roots of the perfect squares (e.g., 0, 1, 4, 9, 16) are integers. In all other cases, the square roots of positive integers are irrational numbers, and hence have non-repeating decimals in their decimal representations.* [2]  <https://en.wikipedia.org/wiki/Square_root>


While getting the square root of the perfect squares is straightforward enough, getting the square root of numbers that are not perfect squares is not so. A manual calculation would start by looking at the square roots of the nearest perfect squares on either side of the number in question and involves some guesswork to begin. The Babylonian method of guess and divide takes a starting estimate to the square root ($x_0$), divides this starting estimate into the number $a$ whose square root you are looking for and then finds the average of the quotient $x_0$ and the divisor $a$ to get the next estimate $x_{0+1}$.

### Calculate the square root of number using Babylonian method
Finding the square root of a number that is not a whole number begins with some guesswork. Begin by looking at the nearest perfect squares on both sides of the number to make a starting guess of what the square root could be.
The Babylonian method of guess and divide takes a starting estimate to the square root ($x_0$), divides this starting estimate into the number $a$ whose square root you are looking for and then finds the average of the quotient $x_0$ and the divisor $a$ to get the next estimate $x_{0+1}$.

The post here at https://blogs.sas.com/content/iml/2016/05/18/newtons-method-babylonian-square-root.html explains simply how the algorithm taught in school for calculating the estimating a square root from an arbitrary guess is an iterative procedure. Given any positive number $S$ and any positive number $x_0$, apply the following formula:
$$x_{n+1} = \frac{(x_n + S/x_n)}{2}$$ until the iteration converges to the square root of $S$.
 
The same author has a post on the [Babylonian method for finding square roots by hand](https://blogs.sas.com/content/iml/2016/05/16/babylonian-square-roots.html). (This seems to be the way I was trying to work out below.)
Suppose you are given any positive number $S$. To find it's square root do the following:
1. Make an initial guess. This guess can be any positive number. This is $x_0$
2. Improve the guess by applying the formula $x_1 = \frac{(x_0 + \frac{S}{x_0})}{2}$.
   The number $x_1$ is a better approximation to the square root of $S$.
3. Iterate until convergence. 
   Apply the formula $x_{n+1}=\frac{(x_n+\frac{S}{x_n})}{2}$ until the process converges.

    * Convergence is achieved when the  digits of $x_{n+1}$ and $x_n$ agree to as many decimal places as you desire.
    
#### Applying this logic to get the square root of 2

First make a guess of what the root of the number in question could be based on the square root of the nearest perfect squares on either side.  Add the result of this to the number and question and divide by two to get the average as the new starting guess. Keep repeating.

1. Find two perfect square roots that the number in question lies between. 
 
2. Take one of those square roots as an estimate and divide the number in question by it.
    * S = 2
    * The square root of 2 must lie between 1 and 2
    * Taking 1 here as the estimate $x_0=1$
    * Divide number in question by the starting estimate S/x0 = 2/1 = 2 = X1
    
3. Take the average of the latest estimate x1 (2) and S (from the previous step above) and the Number.

    * x1 = (x0 +S/x0)/2 = (2 + 2/1)/2 = 2
    * x2 = (x1 + S/x1)/2 = (2 + 2/2)/2 =1.5
    * x3 = (x2 + S/x2)/2 = (1.5 + 2/1.5)/2 = 1.4166666666666665
    * x4 = (x3 + S/x3)/2 = (1.4166666666666665 + 2/1.4166666666666665)/2 = 1.4142156862745097
    * x5 = (x4 + S/x4)/2 = (1.4142156862745097 + 2/1.4142156862745097)/2 = 1.4142135623746899
    * x6 = (x5 + S/x5)/2 = (1.4142135623746899 + 2/1.4142135623746899)/2 = 1.414213562373095
    * x7 = (x6 + S/x6)/2 = (1.414213562373095 + 2/1.414213562373095)/2   = 1.414213562373095
    * x8 = (x7 + S/x7)/2 = (1.414213562373095 + 2/1.414213562373095)/2   = 1.414213562373095
    * x9 = (x8 + S/x8)/2 = (1.414213562373095 + 2/1.414213562373095)/2   = 1.414213562373095
    
    `math.sqrt(10)` gives 3.1622776601683795 

As the illustration shows here, once you get to the 5th iteration, the result is truncuated or rounded. (which?)
- The square root of 2 truncuated to 50 decimal places is 1.41421356237309504880168872420969807856967187537694
See (https://en.wikipedia.org/wiki/Square_root).


In [26]:
#print(f"{(3.16666667 + (10/3.1666667))/2}")
#print(f"{(3.162280686800554 +10/3.162280686800554)/2}")
#print(f"{(3.1622776601698277 + 10/3.1622776601698277)/2}")
#print(f"{(3.162277660168379 + 10/3.162277660168379)/2 }")

print(f"{(1.4166666666666665 + 2/1.4166666666666665)/2}")
print(f"{(1.4142156862745097 + 2/1.4142156862745097)/2}")
print(f"{(1.4142135623746899 + 2/1.4142135623746899)/2}")
print(f"{(1.414213562373095 + 2/1.414213562373095)/2}")
print(f"{(1.414213562373095 + 2/1.414213562373095)/2}")
# and on until you get the required accuracy

1.4142156862745097
1.4142135623746899
1.414213562373095
1.414213562373095
1.414213562373095



### Newton's Method
Newtons Method which is also known as the Newton–Raphson method, can be used to get an estimate of the square root of a number.  It is a root-finding algorithm which produces successively better approximations to the roots (or zeroes) of a real-valued function. [Newton's Method](https://en.wikipedia.org/wiki/Newton%27s_method) is one of many methods of computing square roots and is outlined in detail in the Wikipedia article. It outlines how the problem of finding the square root of a number $a$ can be rephrased as finding the zero of $f(x)=x2−a$ where $f$ is a single-variable function defined for a real variable $x$, with an initial guess of $x_0$ for a root of $f$. The derivative of the function $f'(x)=2x$. Rearranging the formula yields the Babylonian method of finding square roots :

$$x_{n+1}=x_{n}-{\frac{f(x_{n})}{f'(x_{n})}}=x_{n}-{\frac{x_{n}^{2}-a}{2x_{n}}}={\frac {1}{2}}{\biggl(}2x_{n}-{\Bigl(}x_{n}-{\frac{a}{x_{n}}}{\Bigr)}{\biggr)}={\frac {1}{2}}{\Bigl(}x_{n}+{\frac {a}{x_{n}}}{\Bigr )}$$

i.e. the arithmetic mean of the guess, $x_n$ and $\frac{a}{x_n}$


Using Newtons Method, regardless of the starting estimate, once the square of the estimate gets close enough to the actual number itself, then that is considered a good approximation for the square root.

https://math.mit.edu/~stevenj/18.335/newton-sqrt.pdf

### Coding up the square root function using methods above.

Newton's method or the Babylonian method can be put into code using an iterative algorithm.

- $S$ represents the number whose square root is being sought. (This is $a$ in the formula above)
- $x$ represents the changing estimates ($x_n$, $x_0$, $x_1$, $x_{n+1}$ for the estimates at each step). 
- The algorithm takes the average of the estimate  $x_n$ and the number  $S$ divided by the estimate $x_n$. This will give a new estimate for the next iteration $x_{n+1}$.
- The values of the estimate $x$ is updated at each iteration of a while loop representing the $X_n$ and $x_{n+1}$ estimates.

The most digits this function yielded was 15 decimal places. While the algorithm itself will calculate approximately the square root of two (accurate to 15 digits at least), it will not print out 100 decimal places so easily. 
At some stage the result gets rounded and this is the main issue as outlined earlier when dealing with real numbers. Wikipedia records the square root of 2 to 50 decimal places as `1.41421356237309504880168872420969807856967187537694`.
Numpy's `sqrt` function returns the same answer as my function above to 15 decimal places. The square root of 2 using `np.sqrt` to 50 decimal places is 1.41421356237309514547462185873882845044136047363281 which differs from that recorded at Wikipedia.

I did try to increase the precision of the while loop to see if this would yield additional digits but the maximum was still  only 15 decimal places. This is as expected given the background of how computers deal with floating point numbers. 
The[Python docs section on on floating points numbers.](https://docs.python.org/3/tutorial/floatingpoint.html) 
>Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.
According to the Python Docs section on [numeric types](https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex) integers have unlimited precision.

Therefore the next logical step was to apply the function to a very large integer multiple of 2 as this would not have the same limitations as floats and then divide the result by the same integer multiple. The whole number before the decimal point cannot be rounded. In the same way that the square root of 4 is 2, the square root of 400 is 20, and so on. In this way the digits of the square root of two could be produced but with the decimal point in the incorrect place. Some string manipulation then to return the correct result.

Another option was applying the power of 1/2 to get the square root and applying this to a large multiple of 2 but again this returns scientific notation after the 15th decimal place.

Applying either Newtons formula or the equivalent Babylonian guess and divide formula resulted in 1.4142135623730951 being returned as the approximate square root of two (using precision of 0.0000000000001 to stop the while loop.)
Applying the function to get the square root of 2e200 gave the result in scientific notation. Therefore the 100 digits cannot be extracted at this stage.
Nevertheless I compared the results to math.sqrt to see how they fare and the results are the same.
Python only prints a decimal approximation to the true decimal value of the binary approximation stored by the machine. (https://docs.python.org/3/tutorial/floatingpoint.html).  While the printed result might look like the exact value, the actual stored value is the nearest representable binary fraction. While string formatting can be used to produce a limited number of significant digits, you are really just rounding the display of the true machine value. 
>The errors in Python float operations are inherited from the floating-point hardware, and on most machines are on the order of no more than 1 part in `2**53` per operation. That’s more than adequate for most tasks, but you do need to keep in mind that it’s not decimal arithmetic and that every float operation can suffer a new rounding error.

In [113]:
200000000000000000000000000000000000000**(1/2)

1.414213562373095e+19

In [219]:
# S is number to get square root of, S= 2
def sqrtn():
    
    # take the first estimate x to be the number itself
    S=2
    x = S 
    
    # repeat this until the difference between the Estimate Squared minus the number is small enough
    # without setting a stopping value here, the loop will continue for a long time
    while abs(x*x - S) > 0.000000000001:
        # Update the new estimate to be the current estimate minus the Number using Newtons formula
        x -= ((x*x)- S)/ (2 * x)
        # or equivalently the babylonian guess and divide formula
        #x= (x + S/x)/2
   
    
    print(x)
    return x
   
sqrtn()

print("done")

1.4142135623730951
done


In [169]:
# define the function
def sqrt2():
    
    S = 2 * 10**200
    # first estimate
    x = S
    # repeat until there is no difference between the squared estimate gives the square back
    # while abs(x*x -S) > 0.000000000001:
    while abs(x*x -S) > 0:
        # update the new estimate to be the mean of the current estimate and the number divided by the estimate
        
        x= (x + S/x)/2
        
        #x -= ((x*x)- S)/ (2 * x) # this gives 1 extra digit
    return x
        
    
print(sqrt2())

1.414213562373095e+100


In [222]:
# string formatting to get 100 significant digits
format(sqrtn(),'.100g')
format(sqrt2(),'.100g')


1.4142135623730951


'1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100'

In [165]:
import math
math_root2 = math.sqrt(2*(10**200))
print(math_root2)
format(math_root2,'.100g')

1.414213562373095e+100


'1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100'

In [229]:
# define the function
def sqrt2():

    sqrt_string =""
    S = 2 # This is the number we want to get the square root of
    
    S = S*(10**200) # 2 by 10 to power of 200

    x = S
   
    # while the difference between the number and the square of the guess is greater than 0
    while abs(S -(x *x)) > 0.000000000000001:
        x= (x + S/x)/2
    
    str_x = str(x)
    
    for s in str_x:
        sqrt_string += s
    
    print(sqrt_string)
    


sqrt2()

print(f"The square root of two using my function sqrt2() is \n{sqrt2()}")

1.414213562373095e+100
1.414213562373095e+100
The square root of two using my function sqrt2() is 
None


A further bit of research found this post from [geeksforgeeks.com](https://www.geeksforgeeks.org/square-root-of-an-integer/?ref=rp) which uses a binary search function to find the largest integer *i* whose square is less than or equal to the original number. The values of i * i is monotonically increasing, so the problem can be solved using binary search.
 
The code there uses integer division and defines the lower and upper bounds, the midpoint and the answer at each iteration. The loop runs until the lower bound is less than or equal to the upper bound. Then it checks if the midpoint squared is equal to the number and narrows the search to one half of the search space. 

If I apply this function below to get the square root of two it returns 1 as it uses integer division. However if I apply it to a very large multiple of two as outlined earlier, the digits do actually match those from Wikipedia for the square root of two. The task however is not to get the square root of 2e200 to 100 decimal places but to get the square root of two so I attempt to do some string manipulation and convert the digits into a string representing the square root of two. 

(~~When I do perform the conversion from the integer into a string with the decimal point in the right place (for the square root of 2) I actually get the same result as from my own function. This might indicate that my original solution might actually be producing the correct answer after all. Integer division did not work with my earlier code but using lower and upper bounds might reduce the number of iterations required and prevent it from crashing.~~)

The code below adapts the binary search code to calculate the square root of 2e200. It then convert the result into a string and performs some string manipulation by first creating an empty string, then iterating through each digit of the string result, extracting each of the digits after the first one and concatenating them to the string "1." This does recreate the first 100 digits of the square root of 2.

In [5]:
# adapted and extended from code on geekforgeeks contributed by Nikita Tiwari. 
  
# Returns floor of square root of x          
def sqrt2() : 
   
    # set up an empty string to hold the result
    sqrt_string =""
    
    S = 2 * 10**200
    # Base cases is only required for the general case
    if (S == 0 or S == 1) : 
        return S 
   
    # Do Binary Search to find the square root of S 
    lower = 1
    upper = S    
    while (lower <= upper) : 
        x = (lower + upper) // 2
          
        # If x squared is a perfect square then return x
        if (x*x==S) : 
            return x 
              
        # As we need the floor, update the answer when mid*mid is smaller than x, and move closer to getting the square root of S
        if (x*x < S) : 
            lower = x + 1
            ans = x 
              
        else : 
              
            # If mid*mid is greater than x 
            upper = x-1
                          
    #return ans 

    # create a string version of the result
    str_ans = str(ans)
    
    for s in str_ans:
        sqrt_string += s
      
    # concatenate the strings to produce the square root of 2
    sqrt_string = "1."  + sqrt_string[1:]
    #print(f"The square root of 2 to 100 decimal places is \n{sqrt_string}")
      
    # this will return an integer for square root of 2e200
    #return ans   
    
    # instead return the string containing the first 100 digits of square root of 2 
    print(f"The first 100 digits of the square root of 2 are \n{sqrt_string}")
    return sqrt_string
    
    

# call the function to print the first 100 digits of the square root of 2
sqrt2()

The first 100 digits of the square root of 2 are 
1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


'1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727'

This post [10,000 digits of Square Root of 2](https://nerdparadise.com/math/reference/2sqrt10000) on nerdparadise.com has the digits of the square root of 2 to 10,1000 places:
1.
4142135623 7309504880 1688724209 6980785696 7187537694 8073176679 7379907324 7846210703 8850387534 3276415727 

These are the results of the function above:

14142135623 7309504880 1688724209 6980785696 7187537694 8073176679 7379907324 7846210703 8850387534 3276415727



#### Some notes on the Square root of 2 task.

- The final function above does yield the correct digits for the square root of 2. It is a string however and further numerical calculations cannot be performed on the string. Converting from one type to another will not always return results as expected. 
- My original function using the babylonian guess and divide method does return the exact same result as  `math.sqrt` to 15 decimal places. 

- This task illustrates the problems with accuracy to high levels of precision and how when converting between types you won't always get what you expect. The lecture on data types and numbers highlighted that there are some irrational numbers that simply cannot be stored on a computer. Procedures can be written to produce the value when needed though. The article from the Python tutorial on the problems with floating point numbers showed an example of how there are many different decimal numbers which actually share the same nearest approximate binary fraction. It demonstrated how the numbers 0.1 and 0.10000000000000001 and 0.1000000000000000055511151231257827021181583404541015625 are all approximated by 3602879701896397 / 2 ** 55.

In reality, while we would be concerned with the accuracy of results, other considerations would be the computational cost and time.  When I tried to run some versions of my code, my computer struggled!  Newton's method uses approximations and while it converges towards the actual result, the number of steps to reach this is infinite. This is combined with the problems outlined earlier of computing and storing real numbers on a computer.
The post on [Square Roots via Newton’s Method](https://math.mit.edu/~stevenj/18.335/newton-sqrt.pdf) on MIT course 18.335 implements the square root of 2 using the Julia language. It notes that we need to be concerned with the time and computational resources required in addition to the accuracy of the result.

References. I will fix these up and place them at the bottom of the notebook.
Task 1
- [1][Wikipedia](https://en.wikipedia.org/wiki/Square_root)
- [2][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2)
- [3][geeksforgeeks](https://www.geeksforgeeks.org/square-root-of-2/)
- [4][mathsisfun](https://www.mathsisfun.com/numbers/square-root-2-irrational.html)
- [5][The Maths Book: Big Ideas Simply explained](published by Dorling Kindersley Limited)
- [6] Lecture Notes from Machine Learning and Statistics, Lecturer Dr. Ian McLoughlin
The lecture on data types of this module looked at the binary representation of integers and floats.
- [7][Floating Point Arithmetic: Issues and Limitations](https://docs.python.org/3/tutorial/floatingpoint.html) 
- [8][IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) 
- [9][IEEE Standard 754 Floating Point Numbers](https://www.geeksforgeeks.org/ieee-standard-754-floating-point-numbers/) article on <geeksforgeeks.com>.
- [10][The Perils of Floating Point](http://www.lahey.com/float.htm)
- [11][str.format](https://docs.python.org/3/library/stdtypes.html#str.format)
- [12][Python `math` module](https://docs.python.org/3/library/math.html?highlight=math#module-math).

---
# Task 2 
The Chi-squared test for independence is a statistical hypothesis test like a t-test. It is used to analyse whether two categorical variables are independent. The Wikipedia article gives the table below as an example [4], stating the Chi-squared value based on it is approximately 24.6. Use scipy.stats to verify this value and calculate the associated p value. You should include a short note with references justifying your analysis in a markdown cell.

|   | A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|90 | 60 | 104 | 95 | 349 |
|Blue Collar |30 |50 |51|20|151
|No Collar |30 |40|45|35|150
|Total|150 |150|200|150 |650|

- [https://en.wikipedia.org/w/index.php?title=Chi-squaredtest&oldid=983024096](https://en.wikipedia.org/w/index.php?title=Chi-squaredtest&oldid=983024096)

### Extract from Wikipedia Article in question:
Here is an extract from the article in question on the [Chi-squared test](https://en.wikipedia.org/wiki/Chi-squared_test) on Wikipedia from which the above table is taken. 

>A chi-squared test, also written as $\chi^2$ test, is a statistical hypothesis test that is valid to perform when the test statistic is chi-squared distributed under the null hypothesis, specifically Pearson's chi-squared test and variants thereof. Pearson's chi-squared test is used to determine whether there is a statistically significant difference between the expected frequencies and the observed frequencies in one or more categories of a contingency table.
In the standard applications of this test, the observations are classified into mutually exclusive classes. If the null hypothesis that there are no differences between the classes in the population is true, the test statistic computed from the observations follows a $\chi^2$ frequency distribution. The purpose of the test is to evaluate how likely the observed frequencies would be assuming the null hypothesis is true.
Test statistics that follow a $\chi^2$ distribution occur when the observations are independent and normally distributed, which assumptions are often justified under the central limit theorem. There are also $\chi^2$ tests for testing the null hypothesis of independence of a pair of random variables based on observations of the pairs.
Chi-squared tests often refers to tests for which the distribution of the test statistic approaches the $\chi^2$ distribution asymptotically, meaning that the sampling distribution (if the null hypothesis is true) of the test statistic approximates a chi-squared distribution more and more closely as sample sizes increase.

##### Example chi-squared test for categorical data (extract from same Wikipedia article as above)
>Suppose there is a city of 1,000,000 residents with four neighborhoods: A, B, C, and D. A random sample of 650 residents of the city is taken and their occupation is recorded as "white collar", "blue collar", or "no collar". The null hypothesis is that each person's neighborhood of residence is independent of the person's occupational classification. The data are tabulated as:
Let us take the sample living in neighborhood A, 150, to estimate what proportion of the whole 1,000,000 live in neighborhood A. Similarly we take $\frac{349}{650}$ to estimate what proportion of the 1,000,000 are white-collar workers. By the assumption of independence under the hypothesis we should "expect" the number of white-collar workers in neighborhood A to be:
$$150 x \frac{349}{650}\approx80.54$$
Then in that "cell" of the table, we have $\frac{(observed-expected)^2}{expected}$ = $\frac{(90-80.54)^2}{80.54}\approx1.11$.
The sum of these quantities over all of the cells is the test statistic; in this case, $\approx24.6$. Under the null hypothesis, this sum has approximately a chi-squared distribution whose number of degrees of freedom are $\text{number of rows}-1)({\text{number of columns}}-1)=(3-1)(4-1)=6$
If the test statistic is improbably large according to that chi-squared distribution, then one rejects the null hypothesis of independence.

>A related issue is a test of homogeneity. Suppose that instead of giving every resident of each of the four neighborhoods an equal chance of inclusion in the sample, we decide in advance how many residents of each neighborhood to include. Then each resident has the same chance of being chosen as do all residents of the same neighborhood, but residents of different neighborhoods would have different probabilities of being chosen if the four sample sizes are not proportional to the populations of the four neighborhoods. In such a case, we would be testing "homogeneity" rather than "independence". The question is whether the proportions of blue-collar, white-collar, and no-collar workers in the four neighborhoods are the same. However, the test is done in the same way.

### Some notes on the Chi-Squared test
(based on the above article, some previous college notes, and other sources referenced.)

- The Chi-Squared test is used to analyse whether two categorical variables are independent.
Categorical data is data where each observation belongs to one of a set of categories. There are 2 types of categorical data:
    1. Nominal data has two or more categories but there is no ordering of the categories such as Gender - male / female
    2. Ordinal where there is a clear ordering of the variables such as education levels (primary, secondary, tertiary)
Categorical data is often referred to as count data because the number of observations belonging to a particular category are counted.
- If only one categorical variable is measured on each member of the sample, then the data are considered univariate or one-dimensional. 
- If two variables are measured the data are bivariate.
With a sample from a single population, the data are best summarised in a one-way-frequency table (frequency distribution) which lists the possible categories and the number of sample values belonging to each (frequency).
Begin by summarising the sample data in a one-way frequency table with k cells (k is the number of categories). The frequencies are often referred to as cell counts. The hypotheses to be tested concern the probabilities p1, p2, ..., pk which are often proportions of a population falling into each category.


- A $\chi^2$ test is used to determine if there is a statistically significant difference between the expected frquencies and the observed frequencies in one or more categories of a contingency table.
- In the standard applications of this test, the observations are classified into mutually exclusive classes.
 
- If the null hypothesis that there are no differences between the classes in the population is true, the test statistic computed from the observations follows a $\chi^2$ frequency distribution. 
- The purpose of the test is to evaluate how likely the observed frequencies would be assuming the null hypothesis is true.

### Explaining the table data.

- The city has a population of 1,000,000 residents with four neighbourhoods, A, B ,C and D.
- A random sample of 650 residents is taken. 
- The occupation of the 650 sampled residents is recorded as "white collar", "blue collar" or "no collar".

To see if the proportions of residents living in the neighbourhoods (belonging to the categories) 'A', 'B','C' and 'D' are the same for the occupational categories of White Collar, Blue Collar and No Collar, a sample is taken and one-way frequency tables are constructed for each category. Two or more one-way frequency tables can be combined to get a two-way frequency table which is also known as a contingency table.

The table here is a two-way frequency table.

|   | A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|90 | 60 | 104 | 95 | 349 |
|Blue Collar |30 |50 |51|20|151
|No Collar |30 |40|45|35|150
|Total|150 |150|200|150 |650|

- The cell counts are the numbers in a particular sample that belong to a particular category.
- The column and row totals are known as the marginal totals and each set of marginal totals adds to the grand total. 
- In this case the row totals are the sizes of the samples selected from the different occuational populations (i.e. white-collar, blue-collar, no-collar).
- The column totals are the sizes of the samples selected from the different neighbour populations A, B, C and D.
- The null hypothesis is that each person's neighbourhood of residence is independent of the person's occupational classification.


- 150 of the sample live in neighbourhood A. This figure can be used to estimate what proportion of the 1 million residents live in neighbourhood A. 
- 150 of the sample live in neighbourhood B. This figure can be used to estimate what proportion of the 1 million residents live in neighbourhood B.
- 200 of the sample live in neighbourhood C. This figure can be used to estimate what proportion of the 1 million residents live in neighbourhood C.
- 95 of the sample live in neighbourhood D. This figure can be used to estimate what proportion of the 1 million residents live in neighbourhood D.


- In the same way, 349 of the sample are white-collar workers. This figure can be used to determine what proportion if the 1 million residents are white-collar workers.

- 151 of the sample are blue-collar workers. This figure can be used to determine what proportion if the 1 million residents are blue-collar workers.
- 150 of the sample are no-collar workers. This figure can be used to determine what proportion if the 1 million residents are no-collar workers.

- Under the null hypothesis (that each person's neighbourhood of residence is independent of the person's occupational classification) we would expect the number of white collar workers in the neightbourhood *A* to be



#### One-way frequency table
The first row of the table for white collar workers in the sample is a one-way frequency table with $k$ = 4 categories for a sample of 349.

|Observed Cell Counts| A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|90 | 60 | 104 | 95 | 349 |

A question here could be does the data indicate a preference for white collar residents to prefer a particular neighbourhood? If a white-collar worker has no preference for a particular neighbourhood then the proportion of the population in each of the four neighbourhoods would be $\frac{1}{4}$. The hypothesis would be $H_0: p_1 = p_2 = p_3 = p_4 = \frac{1}{4}$ vs $H_A: H_0: \text{not true}$ (or at least one proportion is not $\frac{1}{4}$ ).

The null hypothesis is that the white collar workers have no preference for a particular neighbourhood (i.e. the proportions for each of the categories are equal) and the alternative hypothesis is that the null is not true (i.e. the proportions for each of the categories are not the same). To test the hypotheses you need to calculate the frequencies (cell counts) that would be expected if the null hypothesis was true, ie if the white-collar workers had no preference for a particular neighbourhood. These expected cell counts are $np_1$, $np_2$ etc.


The table above shows the observed counts from the experiment.
If the null hypothesis was true, that the white-collar workers have no neighbourhood preference, we would expect to see the following:

|ExpectedCell Counts| A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|87.25| 87.25 | 87.25 | 87.25 | 349 |

    - $np_1$ = 349 * 0.25
    - $np_2$ = 349 * 0.25
    - $np_3$ = 349 * 0.25
    - $np_4$ = 349 * 0.25

The next step is to compare the observed and the expected cell counts. 
If the differences between them are so big that they could not be explained by sampling variation then we reject the null hypothesis in favour of the alternative hypothesis. The comparison of observed and expected cell counts is made by calculating:


$$\frac{(Observed Count - Expected Count)^2}{Expected Count}$$ for each cell and then adding the values for all cells to get the chi-square test statistics.

$$\chi^2=\sum_{i=1}^{k}\frac{(O_i-E_i)^2}{E_i}$$

- Large values of $\chi^2$ suggest rejection of the Null hypothesis in favour of the alternative hypothesis. 
- Small values of $\chi^2$ would support the null hypothesis.

In [220]:
A, B, C, D =90, 60, 104,95

Total = A+B+C+D
# Actual observed proportions
print(f"Total: {Total}, Proportion A: {A/Total}, Proportion B: {B/Total},Proportion C: {C/Total}, Proportion D: {D/Total}")
# Expected Proportions under the null hypothesis
print(f"Total {Total}, \nExpected A: {Total *0.25}\nExpected B: {Total *0.25}\nExpected C: {Total *0.25}\nExpected D: {Total *0.25}\n ")
#print(f"Total: {Total}, Proportion White Collar: {White/Total}, Proportion Blue Collar: {Blue/Total},Proportion No Collar: {No/Total}, Proportion D: {D/Total}")

Total: 349, Proportion A: 0.25787965616045844, Proportion B: 0.17191977077363896,Proportion C: 0.2979942693409742, Proportion D: 0.2722063037249284
Total 349, 
Expected A: 87.25
Expected B: 87.25
Expected C: 87.25
Expected D: 87.25
 




|   | A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|90 | 60 | 104 | 95 | 349 |
|Blue Collar |30 |50 |51|20|151
|No Collar |30 |40|45|35|150
|Total|150 |150|200|150 |650|

The null hypothesis is that each person's neighborhood of residence is independent of the person's occupational classification.
This would mean that the two variables neighbourhood and occupational classification are independent. Knowing a person's occupational classification does not tell you anything about where they live.
This considers whether the proportion of workers living  in the neighbourhoods "A", "B","C" and "D" (belonging to the categories) are the same for the person's classification as White-Collar, Blue-Collar or No-Collar. 

Select a sample and construct a one-way frequency table for each classification.

The table provided in Wikipedia is a two-way frequency table. It is comparing more than two populations. 

To see if the proportions of persons belonging to the categories 'A', 'B', C' and 'large' are the same for the occupational classifications of White-collar, Blue-collar and No-Collar, a sample is selected and a one-way select one-way frequency table is constructed for each occupational category. 
A two-way table, also known as a contingency table, is the result of combining two or more one-way frequency tables.
The cell counts are the numbers in a particular sample that belong to a particular category. The column and row totals are known as the marginal totals and each set of marginal totals adds to the grand total. In this case the row totals are the sizes of the samples selected from the different populations (i.e. occupational classification of residents).
The Null Hypothesis would be that the neighbourhood category proportions are the same for all occupational classifications
- $H_0$ : Neighbourhood category proportions are the same for all classifications of occupations.
- $H_A: H_0$ is not true.
Assume the null hypothesis is true, The Expected cell counts = (row total times the column total) / grand total)
Assuming the null hypothesis to be true $\chi^2$ is approximately chi-squared distributed with df = (Number of rows- 1).(Number of columns - 1). In this case the degrees of freedom (3-1 x 4-1) is 6. If calculated manually the test statistic would be compared to the critical value (from a table such as the New Cambridge Statistical Tables) at a particular alpha level such as 0.05 taking into account the degrees of freedom. 

- If the chi-squared test statistic is less than the chi-squared critical value with 6 degrees of freedom (at the chosen alpha level such as 0.05) then you would fail to reject the null hypothesis.
- If the chi-squared test statistic is greater than the chi-squared critical value, then the null hypothesis would be rejected and you would accept the alternative hypothesis that the data contains sufficient evidence of neighbourhood preference by occupational classification. The proportions of residents in the four neighbourhoods A, B, C and D are not the same for the 3 occupational classifications of white-collar, blue-collar and no-collar.

#### Proportions of the sample

In [249]:
A, B, C, D =150, 150, 200,150
White, Blue, No = 349,151,150
Total = A+B+C+D
print(f"Total: {Total}, Proportion A: {A/Total}, Proportion B: {B/Total},Proportion C: {C/Total}, Proportion D: {D/Total}")
print(f"Total: {Total}, Proportion White Collar: {White/Total}, Proportion Blue Collar: {Blue/Total},Proportion No Collar: {No/Total}, Proportion D: {D/Total}")

Total: 650, Proportion A: 0.23076923076923078, Proportion B: 0.23076923076923078,Proportion C: 0.3076923076923077, Proportion D: 0.23076923076923078
Total: 650, Proportion White Collar: 0.536923076923077, Proportion Blue Collar: 0.2323076923076923,Proportion No Collar: 0.23076923076923078, Proportion D: 0.23076923076923078


By the assumption of independence under the hypothesus

- $H_0$ : Category proportions are the same for all categories
- $H_A: H_0$ is not true.
- Assume the null hypothesis is true, The Expected cell counts = (row total times the column total) / grand total)

Here I will put the data from the contingency table into a pandas dataframe and then create an expected frequency table based on the observed data.

In [250]:
import numpy as np
import pandas as pd


In [252]:
# create a pandas dataframe
df=pd.DataFrame({"A":[90,30,30], "B":[60,50,40],"C":[104,51,45],"D":[95,20,35]})
df.index = ['White_collar', 'Blue_collar', 'No-Collar'] 
# add a total column to sum across each row
df["Total"] = df.sum(axis=1)
# add a row to sum across column totals 
#(https://stackoverflow.com/questions/20804673/appending-column-totals-to-a-pandas-dataframe)
df.loc['Total']= df.sum()
df

A    B    C    D  Total
White_collar   90   60  104   95    349
Blue_collar    30   50   51   20    151
No-Collar      30   40   45   35    150
Total         150  150  200  150    650


- The column and row totals are known as the **marginal totals**
- Each set of marginal totals adds to the **grand Total**
- in this case the row totals are the sizes of the samples from the different populations (occupational classification)
- Ho : Category proportions in the different neightbourhoods are the same for all occupational classifications 
- HA: H0 not true.
Assume the null hypothesis is true,
The expected cell counts = $\frac{\text{row total} \times \text{column total}}{Grand Total}$


In [254]:
# get the marginal row totals
WhiteCollarTotal =df.iat[0, 4] 
BlueCollarTotal=df.iat[1,4]
NoCollarTotal=df.iat[2,4]
# get the marginal column total
ATotal=df.iat[3,0]
BTotal=df.iat[3,1]
CTotal=df.iat[3,2]
DTotal=df.iat[3,3]
# get the grand total
GrandTotal = df.iat[3,4]

print(GrandTotal)


650


Create a dataframe representing the expected frequencies. The marginal row and column totals as well as the overall grand total are the same as totals for the observed frequencies tables.
Note some rounding is introduced here.
Then calculate the chi-squared test statistic. 
Note this is just for some background on the table and to verify the results from the scipy-stats package which has a functions that will actually do all of this. 

In [256]:
# create a pandas dataframe using dictionaries containing zero values to start
dfE =pd.DataFrame({"A":[0,0,0], "B":[0,0,0],"C":[0,0,0],"D":[0,0,0]})
# add an index
dfE.index = ['White_collar', 'Blue_collar', 'No-Collar'] 

# Calculate the White collar expected frequencies
dfE.iat[0,0]=round((WhiteCollarTotal*ATotal)/GrandTotal)
dfE.iat[0,1]=round((WhiteCollarTotal*BTotal)/GrandTotal)
dfE.iat[0,2]=round((WhiteCollarTotal*CTotal)/GrandTotal)
dfE.iat[0,3]=round((WhiteCollarTotal*DTotal)/GrandTotal)
# Blue collar
dfE.iat[1,0]=round((BlueCollarTotal*ATotal)/GrandTotal)
dfE.iat[1,1]=round((BlueCollarTotal*BTotal)/GrandTotal)
dfE.iat[1,2]=round((BlueCollarTotal*CTotal)/GrandTotal)
dfE.iat[1,3]=round((BlueCollarTotal*DTotal)/GrandTotal)
# No Collar
dfE.iat[2,0]=round((NoCollarTotal*ATotal)/GrandTotal)
dfE.iat[2,1]=round((NoCollarTotal*BTotal)/GrandTotal)
dfE.iat[2,2]=round((NoCollarTotal*CTotal)/GrandTotal)
dfE.iat[2,3]=round((NoCollarTotal*DTotal)/GrandTotal)

# add a total column to sum across each row
dfE["Total"] = dfE.sum(axis=1)
# add a row to sum across column totals 
dfE.loc['Total']= dfE.sum()
dfE

A    B    C    D  Total
White_collar   81   81  107   81    350
Blue_collar    35   35   46   35    151
No-Collar      35   35   46   35    151
Total         151  151  199  151    652

In [232]:
# calculate the chi-squared statistic using the formula
chi = ((df-dfE)**2)/dfE
chi
chi["Total"] = chi.sum(axis=1)

chi.loc['Total']= chi.sum()
chi

A          B         C         D      Total
White_collar  1.000000   5.444444  0.084112  2.419753   8.951167
Blue_collar   0.714286   6.428571  0.543478  6.428571  14.114907
No-Collar     0.714286   0.714286  0.021739  0.000000   1.456933
Total         2.435194  12.593924  0.654355  8.854947  24.554034

I need to do some rounding here but the value is close to the Chi-squared value in the Wikipedia article.
The next step is to use the `scipy.stats` to verify this value and calculate the associated p value. 
Include a short note with references justifying your analysis in a markdown cell.


|   | A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|90 | 60 | 104 | 95 | 349 |
|Blue Collar |30 |50 |51|20|151
|No Collar |30 |40|45|35|150
|Total|150 |150|200|150 |650|

## Scipy.stats functions for Chi-Squared tests

- [Scipy stats - Statistical Tests](https://docs.scipy.org/doc/scipy/reference/stats.html#statistical-tests)

- `scipy.stats.chisquare(f_obs, f_exp=None, ddof=0, axis=0)`
- [Calculate a one-way chi-square test.](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html#scipy.stats.chisquare)
The chi-square test tests the null hypothesis that the categorical data has the given frequencies.
- The function takes as parameters an array of observed frequencies in each category, an array of expected frequencies in each category and the degrees of freedom. By default the categories are assumed to be equally likely.
The expected frequencies for the table from Wikipedia are not provided in the article but can be calculated.

The [Scipy.stats.chi2_contingnency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html#scipy.stats.chi2_contingency) function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the *observed* contingency table.  The expected frequencies are computed based on the marginal sums under the assumption of independence. The [scipy.stats.contingency.expected_freq(observed)](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.contingency.expected_freq.html#scipy.stats.contingency.expected_freq) function can be used to compute the expected frequencies from a contingency table.
Given an n-dimensional contingency table of observed frequencies, compute the expected frequencies for the table based on the marginal sums under the assumption that the groups associated with each dimension are independent.
The parameters this function takes is an array of observed values. The function returns an array of expected frequencies, based on the marginal sums of the table. The expected values returned will be of type `float`. The shape of the expected values will be the same as the observed.
So for example, our table has 3 rows and 4 columns of observed values. Therefore the array of expected values will be the same. The expected values are calculated in the same way as I showed above behind the scenes using marginal sums of the table.


In [257]:
# observed is the multidimensional arrays of values from the original table. 
observed = np.array([[90,60,104,95],[30,50,51,20],[30,40,45,35]])
observed


array([[ 90,  60, 104,  95],
       [ 30,  50,  51,  20],
       [ 30,  40,  45,  35]])

#### Calculating expected frequencies  using Scipy

In [258]:
from scipy.stats.contingency import expected_freq
expected_freq(observed)

array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
       [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
       [ 34.61538462,  34.61538462,  46.15384615,  34.61538462]])

In [27]:
observed.shape

(3, 4)

The [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) is used to calculate the Chi-square test of independence of variables in a contingency table and can be applied to the table of data here.

This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table observed. The number of degrees of freedom is (expressed using numpy functions and attributes).
The function takes as parameters the contingency table which contains the observed frequencies  or number of occurences in each category. (In the two-dimensional case, the table is often described as an “R x C table”.)

[scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html#scipy.stats.chi2_contingency)
- `scipy.stats.chi2_contingency(observed, correction=True, lambda_=None)`

Chi-square test of independence of variables  on a contingency table.


|   | A | B | C | D | Total|
| -- | -- | -- | -- | -- | -- |
|White Collar|90 | 60 | 104 | 95 | 349 |
|Blue Collar |30 |50 |51|20|151
|No Collar |30 |40|45|35|150
|Total|150 |150|200|150 |650|

The chi-squared test statistic is a test for the independence of different categories of a population. It should only be used when the dimension of observed is two or more. 

This function computes a chi-square statistic. It computes the expected frequencies and the degrees of freedom from the given contingency table. (The degrees of freedom is calculated as follows:
`dof = observed.size - sum(observed.shape) + observed.ndim - 1`).

The function returns the chi-squared test statistic as a float, the p-value of the test (also a float), the degrees of freedom (an integer) and the expected frequencies based on the marginal sum of the table of observed frequencies (an ndarray of the same shape as the table of observed).

In [29]:
dof = observed.size - sum(observed.shape) + observed.ndim - 1
dof

6

In [260]:
# import from scipy.stats
from scipy.stats import chi2_contingency
# The array of observed frequencies
obs = np.array([[90,60,104,95], [30,50,51,20],[30,40,45,35]])
obs

array([[ 90,  60, 104,  95],
       [ 30,  50,  51,  20],
       [ 30,  40,  45,  35]])

In [261]:
chi2_contingency(obs)
#print(f"The chi-squared statistic calculated using the scipy stats package is {chi2_contingency(obs)[0]} ")

(24.5712028585826,
 0.0004098425861096696,
 6,
 array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
        [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
        [ 34.61538462,  34.61538462,  46.15384615,  34.61538462]]))

In [264]:
print(f"The chi-squared statistic calculated using the scipy stats package is {chi2_contingency(obs)[0]} ")
print(f"The p-value is {chi2_contingency(obs)[1]} ")

The chi-squared statistic calculated using the scipy stats package is 24.5712028585826 
The p-value is 0.0004098425861096696 


The chi-squared value is compared to the chi-squared critical value which takes into account the alpha level of significance and the degrees of freedom. If the chi-squared test statistic is greater than the critical value of chi at that alpha level for the number of degrees of freedom, then the null hypothesis of independence is rejected and we can conclude that the data contains sufficient evidence of neighbourhood preference by occupational categories.
The p-value is calculated here using a significance level of 0.05 for the hypothesis test. The calculated p-value is 0.0004098425861096696 and this is less than the alpha level of 0.05 which means that the null hypothesis should be rejected in favour of the alternative hypothesis. We can conclude that the data supports the hypothesis that the choice of neighbourhood is not independent of occupational classification. The categorical variables in this data are not independent.
When the observed and expected cell counts are compared, if the differences between them are too big to be explained by sampling variation then the null hypothesis is rejected in favour of the alternative hypothesis. Large values of the chi-squared suggest rejection of the null hypothesis in favour of the alternative hypothesis. The overall difference between what was observed and what would have been expected under the nul hypothesis is larger. Smaller values of chi-squared test stat are compatible with the null hypothesis. The null hypothesis is rejected if the chi-squared test statistic is greater than some critical value. The critical value is chosen in such as way that the probability of a type 1 error is equal to alpha. To do this you need to know the sampling distribution of chi-squared when the null hypothesis is true. The distribution is approximately a chi-squared distribution with k-1 degrees of freedom. The conditions of Chi-squared is that the  approximation is good  when every expected cell count is greater than or equal to 5, and the n observed counts are a random sample from the population of interest.
The critical value for a chi-squared test can be looked up in Table 8 of the New Cambridge Statistical tables. You need the alpha level and the degrees of freedom.
For this example, alpha = 0.05 and degrees of freedom = k-1 = 6. The chi-squared critical value from page 41 of the tables is 12.59. I will look this up online also. Since the chi-squared test statistic is greater than the critical value we can reject the null hypothesis and conclude that the data contains sufficient evidence of neighbourhood preference by occupational classification.

The chi-squared statistic calculated here by scipy is the same as I manually calculated above and also the same as reported in the Wikipedia article referred to in the task description.

# References for Task 2


- Wikipedia contributors, “Chi-squared test — Wikipedia, the free encyclopedia,” 2020, [Online; accessed 1-November-2020][Online]. Available: https://en.wikipedia. org/w/index.php?title=Chi-squared test&oldid=983024096

---
# Task 3
November 16th, 2020: The standard deviation of an array of numbers x is calculated using numpy as `np.sqrt(np.sum((x - np.mean(x))**2)/len(x))` . However, Microsoft Excel has two different versions of the standard deviation calculation, `STDDEV.P` and `STDDEV.S`.
The `STDDEV.P` function performs the above calculation but in the `STDDEV.S` calculation the division is by `len(x)-1` rather than `len(x)`. 
- Research these Excel functions, writing a note in a Markdown cell about the difference between them. Then use numpy to perform a simulation demonstrating that the `STDDEV.S` calculation is a better estimate for the standard deviation of a population when performed on a sample. 
Note that part of this task is to figure out the terminology in the previous sentence

First I will look up the Microsoft documentation first.
The standard deviation is a measure of how widely values are dispersed from the average value (the mean).

The [STDEV function](https://support.microsoft.com/en-us/office/stdev-function-51fecaaa-231e-4bbb-9230-33650a72c9b0) estimates standard deviation based on a sample. According to the documentation, this function has been replaced with one or more new functions that may provide improved accuracy and whose names better reflect their usage.
- The [STDDEV.S function](https://support.microsoft.com/en-us/office/stdev-s-function-7d69cf97-0c1f-4acf-be27-f3e83904cc23) estimates standard deviation based on a sample (ignores logical values and text in the sample).
`STDEV.S` assumes that its arguments are a sample of the population. The standard deviation is calculated using the "n-1" method. If your data represents the entire population, then compute the standard deviation using `STDEV.P`.



The aim of statistics is to make statements about the population using a sample. Therefore it is important to establish the relationship between the sample and the population.
A parameter is a population characterestic while a sample statistic is any quantity computed from  values in the sample. The sample is just one of many that can be drawn from a population. While the value of a population parameter is fixed, the value of a statistic varies from sample to sample. It is a random variable and it's probability distribution known as its sampling distribution.

The standard deviation is the most commonly used measure of the dispersion of a data set. It is the square root of the variance. The larger the standard deviation, the greater the spread of the data. The variance is the average of the squared difference (of each data point) from the mean.
$n-1$ is used in the sample variance formula instead of n because using n tends to produce an underestimate of the population variance. For this reason n-1 is used to provide the appropriate correction for this tendency.


The population variance is calculated using the formula:
$$\sigma^2\frac{=\sum(x_i-\mu)^2}{n}$$

The sample variance is calculated using the formula:

$$S^2\frac{=\sum(x_i-\bar{x})^2}{n-1}$$


According to wikipedia,  using $n−1$ instead of $n$ in the formula for the sample variance and sample standard deviation is known as [Bessel's Correction](https://en.wikipedia.org/wiki/Bessel%27s_correction). The technique is named after Friedrick Bessel. $n$ refers to the number of observations in a sample.
>This method corrects the bias in the estimation of the population variance. It also partially corrects the bias in the estimation of the population standard deviation. However, the correction often increases the mean squared error in these estimations.
>In estimating the population variance from a sample when the population mean is unknown, the uncorrected sample variance is the mean of the squares of deviations of sample values from the sample mean (i.e. using a multiplicative factor 1/n). In this case, the sample variance is a biased estimator of the population variance.
Multiplying the uncorrected sample variance by the factor $\frac{n}{n-1}$ gives an unbiased estimator of the population variance


- Bessel's correction is the degrees of freedom in the residuals vector 
$x_1-\bar{x},...,x_n - \bar{x}$ where $\bar{x}$ is the sample mean. There are $n$ independent observations in the sample but only $n-1$ residuals because they sum to one.

Generally Bessel's correction is an approach to reduce the bias due to finite sample size.

---
If you have a population of size N and take a sample of n from the population. 
The population mean takes the sum of every data point in the population and divide by the number of observations ($n$ data points) to get the population mean which is denoted by $\mu$.
The sample mean is the sum of the observations in a sample divided by the number of observations. This is usually denoted by $\bar{x}$
You then consider the dispersion of the data points from the mean. 
The mean of the squared distances from the population mean divided by the total number of data points in the population.
There are several ways to calculate the sample variance. The **biased** sample variance is a non-unbiased estimator of the population variance and is usually denoted by $S_n$. 

From every one of the $n$ data points in the sample you substract the sample mean $\bar{x}$ and divide by the number of data points $n$. To get the **unbiased** estimate we instead divide by $n-1$. As $n-1$ is a smaller number than $n$, then dividing by the smaller number will give a larger value. This unbiased estimate is usually denoted by $S_{n-1}$.

(Often assume that when people refer to $S^2$ that they are referring to the unbiased estimate).

An example of this would be taking a all the data in a population plotted on the number line. If a sample is taken from this then the sample mean of this sample could be very close to the population mean or then again it might not, depending on what sample of data points is taken. When you take a sample, its sample mean value will always fall inside the sample but the population mean could be outside of this. Therefore you could have a much lower estimate than the true variance from the actual population mean. Depending on what the sample values are and if they come from one end or other of the population, then the true population mean might be outside of the sample taken and then you are likely to be underestimating the true population variance.

If we had all the data in a population plotted on a number line, if a sample is taken from this.
N = 14. take a sample n=3. depending on which 3 points in the sample, the sample mean could be very close to the population mean. There is also a distinct possibility that it won't. The key idea is that when you take a sample, the sample mean will always sit within your sample so there is a real possibility that when you take your sample, the mean could be outside this sample. It could be a much lower estimate than the true variance from the actual population mean. (The sample mean will always sit inside the data but the sample 
Therefore dividing by $n-1$ instead of $n$ will give a slightly larger sample variance and this  will be an unbiased estimate.




### Simulation using Numpy to demonstrate.
The tasks requires using numpy to perform a simulation demonstrating that the `STDDEV.S` calculation is a better estimate for the standard deviation of a population when performed on a sample. 

`np.sqrt(np.sum((x - np.mean(x))**2)/len(x))`

$$frac{\sqrt{\sum{x=\bar{x}}}{n}$$

The variance calculation for the population uses $n$ while the variance calculation for the sample using $n-1$.
The population variance is calculated using the formula:
$$\sigma^2\frac{=\sum(x_i-\mu)^2}{n}$$

The sample variance is calculated using the formula:

$$S^2\frac{=\sum(x_i-\bar{x})^2}{n-1}$$

The standard deviation is the square root of the variance.

So I need to simulate a population using numpy random, then decide on the size of the sample, then from the population select some samples of and calculate the statistics of the samples. Calculate the sample variance of each of the samples using both $n$ and $n-1$ and then take the mean of the variances calculated using $n$ and $n-1$. We should expect to see that when using $n$ instead of $n-1$ that the true variance is still being underestimated. When using $n-1$ the true variance should be closer to the true population variance,

### Task 3 References

- [Wikipedia on Bessels Correction](https://en.wikipedia.org/wiki/Bessel%27s_correction)
- Some old statistics notes from UCD
- [Khan Academy]

# References

More reference above to bring down and include here
- https://www.latex-tutorial.com/tutorials/amsmath/
- https://www.math-linux.com/latex-26/faq/latex-faq/article/latex-natural-numbers
- https://en.wikibooks.org/wiki/LaTeX/Mathematics

Read https://kodlogs.com/blog/715/convert-scientific-notation-to-decimal-python
